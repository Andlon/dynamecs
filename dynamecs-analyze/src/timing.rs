use std::cmp::max;
use std::collections::{HashMap};
use std::time::Duration;
use eyre::eyre;
use time::OffsetDateTime;
use RecordKind::{SpanEnter, SpanExit};
use crate::{Record, RecordKind, SpanPath, SpanTree, SpanTreeNode};
use std::fmt::Write;
use std::iter;

pub type TimingTree = SpanTree<DerivedStats>;
type TimingTreeNode<'a> = SpanTreeNode<'a, DerivedStats>;

/// Statistics measured directly from logs.
#[derive(Debug, Clone, Default)]
pub struct DirectStats {
    /// Total accumulated duration for the span.
    pub duration: Duration,
    /// Number of times the span was entered and subsequently *exited*.
    pub count: u64,
}

impl DirectStats {
    pub fn from_single_duration(duration: Duration) -> Self {
        Self { duration, count: 1 }
    }

    pub fn combine_mut(&mut self, other: &DirectStats) {
        self.duration += other.duration;
        self.count += other.count;
    }
}

#[derive(Debug, Clone)]
pub struct DerivedStats {
    pub duration: Duration,
    pub count: u64,
    pub duration_relative_to_parent: Option<f64>,
    pub duration_relative_to_root: Option<f64>,
}

fn update_column_widths_for_line(column_widths: &mut Vec<usize>, line: &str) {
    let mut column_iter = line.split("\t");
    // Update existing column widths
    for (col_width, column_content) in column_widths.iter_mut().zip(column_iter.by_ref()) {
        *col_width = max(*col_width, column_content.len());
    }
    // Push new column widths
    for column_content in column_iter {
        column_widths.push(column_content.len());
    }
}

fn write_table_line(
    output: &mut String,
    line: &str,
    column_widths: &[usize],
    alignments: &[Alignment],
) {
    let padding = 2;
    debug_assert_eq!(line.lines().count(), 1, "line string must consist of a single line");
    let alignment_iter = alignments.iter().chain(iter::repeat(&Alignment::Left));
    for ((cell, width), alignment) in line.split("\t").zip(column_widths).zip(alignment_iter) {
        match alignment {
            Alignment::Left => write!(output, "{cell:width$}", width=width).unwrap(),
            Alignment::Right => write!(output, "{cell: >width$}", width=width).unwrap()
        }
        for _ in 0 .. padding {
            output.push(' ');
        }
    }
    writeln!(output).unwrap();
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Alignment {
    Left,
    Right
}

fn format_table(header: &str, table: &str, alignments: &[Alignment]) -> String {
    debug_assert_eq!(header.lines().count(), 1, "Header must only have a single line");
    let mut column_widths = vec![];
    update_column_widths_for_line(&mut column_widths, header);
    for line in table.lines() {
        update_column_widths_for_line(&mut column_widths, line);
    }

    let mut output = String::new();
    // Use default alignment for table headers, apply alignments only to cells
    write_table_line(&mut output, header, &column_widths, &[]);
    let header_len = output.len();
    output.push_str(&"═".repeat(header_len));
    writeln!(output).unwrap();

    for line in table.lines() {
        write_table_line(&mut output, line, &column_widths, alignments);
    }

    output.push_str(&"═".repeat(header_len));
    writeln!(output).unwrap();

    output
}

pub fn format_timing_tree(tree: &TimingTree) -> String {
    let mut table = String::new();
    write_timing_tree_node(&mut table, tree.root(), &mut vec![]);
    use Alignment::{Left, Right};
    format_table("Total\tAverage\tCount\tRel parent\tRel root\tSpan",
                 &table,
                 &vec![Right, Right, Right, Right, Right, Left])
}

fn write_proportion(output: &mut String, proportion: Option<f64>) {
    if let Some(proportion) = proportion {
        let percentage = 100.0 * proportion;
        let _ = write!(output, "{percentage:5.1} %");
    } else {
        let _ = write!(output, "    N/A");
    }
}

fn write_timing_tree_node(
    output: &mut String,
    node: TimingTreeNode,
    active_stack: &mut Vec<bool>
) {
    let duration = node.payload().duration;
    let count = node.payload().count;
    write_duration(output, &duration);
    write!(output, "\t").unwrap();
    write_duration(output, &duration.div_f64(count as f64));

    write!(output, "\t{count}").unwrap();

    write!(output, "\t").unwrap();
    write_proportion(output, node.payload().duration_relative_to_parent);
    write!(output, "\t").unwrap();
    write_proportion(output, node.payload().duration_relative_to_root);

    write!(output, "\t").unwrap();
    if let Some((&parent_is_active, predecessors)) = active_stack.split_last() {
        for &is_active in predecessors {
            if is_active {
                output.push_str("│   ");
            } else {
                output.push_str("    ");
            }
        }
        if parent_is_active {
            output.push_str("├── ");
        } else {
            output.push_str("└── ");
        }
    }

    writeln!(output, "{}", node.path().span_name().unwrap_or("no name? Fix, TODO")).unwrap();
    let num_children = node.count_children();
    for (child_idx, child) in node.visit_children().enumerate() {
        // We say that an ancestor is "active" if it's not yet processing its last child.
        // This criterion lets us avoid drawing excessive numbers of vertical lines,
        // which make for a visually confusing picture.
        let is_last_child = child_idx + 1 == num_children;
        active_stack.push(!is_last_child);
        write_timing_tree_node(output, child, &mut *active_stack);
        active_stack.pop();
    }
}

// TODO: Unit tests for this one?
fn write_duration(output: &mut String, duration: &Duration) {
    let secs = duration.as_secs_f64();
    if 1e-9 <= secs && secs < 1e-6 {
        write!(output, "{:5.1} ns", secs / 1e-9).unwrap();
    } else if 1e-6 <= secs && secs < 1e-3 {
        write!(output, "{:5.1} μs", secs / 1e-6).unwrap();
    } else if 1e-3 <= secs && secs < 1.0 {
        write!(output, "{:5.1} ms", secs / 1e-3).unwrap();
    } else if 1.0 <= secs && secs < 1e3 {
        write!(output, "{:5.1} s ", secs).unwrap();
    } else {
        write!(output, "{:5.1e} s ", secs).unwrap();
    }
}

#[derive(Debug, Clone)]
pub struct AccumulatedTimings {
    span_stats: HashMap<SpanPath, DirectStats>,
}

#[derive(Debug, Clone)]
pub struct AccumulatedStepTimings {
    pub timings: AccumulatedTimings,
    pub step_index: u64,
}

impl AccumulatedTimings {
    pub fn new() -> Self {
        Self { span_stats: Default::default() }
    }

    pub fn merge_with_others<'a>(&mut self, others: impl Iterator<Item=&'a AccumulatedTimings>) {
        for other in others {
            for (path, stats) in &other.span_stats {
                let current_stats = self.span_stats.entry(path.clone())
                    .or_default();
                current_stats.combine_mut(&stats);
            }
        }
    }
}

impl AccumulatedTimings {
    pub fn create_timing_tree(&self) -> TimingTree {
        let (paths, durations) = self.span_stats
            .iter()
            .map(|(path, duration)| (path.clone(), duration.clone()))
            .unzip();
        SpanTree::from_paths_and_payloads(paths, durations)
            .transform_payloads(|node| {
                let stats = node.payload();
                let duration = stats.duration;
                DerivedStats {
                    duration: stats.duration,
                    count: stats.count,
                    duration_relative_to_parent: node.parent()
                        .map(|parent_node| {
                            let parent_duration = parent_node.payload().duration;
                            let proportion = duration.as_secs_f64() / parent_duration.as_secs_f64();
                            proportion
                        }),
                    duration_relative_to_root: Some(node.root())
                        // TODO: There will always be a root at the moment,\
                        // but we'll probably later have root nodes with an optional Duration"
                        .map(|root_node| {
                            let root_duration = root_node.payload().duration;
                            let proportion = duration.as_secs_f64() / root_duration.as_secs_f64();
                            proportion
                        })
                }
            })
    }
}

#[derive(Debug, Clone)]
pub struct AccumulatedTimingSeries {
    steps: Vec<AccumulatedStepTimings>,
    // TODO: Timing from other sources outside of steps?
}

impl AccumulatedTimingSeries {
    pub fn summarize(&self) -> AccumulatedTimings {
        let mut summary = AccumulatedTimings::new();
        summary.merge_with_others(self.steps().iter().map(|step| &step.timings));
        summary
    }
}

impl AccumulatedTimingSeries {
    pub fn steps(&self) -> &[AccumulatedStepTimings] {
        &self.steps
    }
}

pub fn extract_step_timings<'a>(records: impl IntoIterator<Item=Record>) -> eyre::Result<AccumulatedTimingSeries> {
    // TODO: Collect statistics from spans outside run as well
    find_and_visit_dynamecs_run_span(records.into_iter())
}

pub fn extract_timing_summary<'a>(records: impl IntoIterator<Item=Record>) -> eyre::Result<AccumulatedTimings> {
    extract_step_timings(records).map(|series| series.summarize())
}

fn find_and_visit_dynamecs_run_span<'a>(mut records: impl Iterator<Item=Record>) -> eyre::Result<AccumulatedTimingSeries> {
    // First try to find the `run` span in the records
    while let Some(record) = records.next() {
        if let Some(span) = record.span() {
            if span.name() == "run"
                && record.target() == "dynamecs_app"
                && record.kind() == RecordKind::SpanEnter {
                return visit_dynamecs_run_span(&record, records);
            }
        }
    }

    Err(eyre!("Could not find new event for `run` span of dynamecs among records"))
}

fn visit_dynamecs_run_span<'a>(run_new_record: &Record, remaining_records: impl Iterator<Item=Record>) -> eyre::Result<AccumulatedTimingSeries> {
    let run_thread = run_new_record.thread_id();
    let run_span_path = run_new_record.span_path();
    let mut iter = remaining_records;

    let mut steps = Vec::new();

    while let Some(record) = iter.next() {
        if record.thread_id() == run_thread {
            if let Some(span) = record.span() {
                match (span.name(), record.target(), record.kind()) {
                    ("step", "dynamecs_app", SpanEnter) => {
                        if let Some(step) = visit_dynamecs_step_span(&record, &mut iter)? {
                            // Only collect complete time steps
                            steps.push(step);
                        }
                    },
                    ("run", "dynamecs_app", SpanExit) if record.span_path() == run_span_path => {
                        break;
                    },
                    // TODO: Still accumulate timings for other things?
                    _ => {}
                }
            }
        }
    }

    Ok(AccumulatedTimingSeries {
        steps
    })
}

/// Returns accumulated timings for the next *complete* step in the records.
fn visit_dynamecs_step_span<'a>(
    step_new_record: &Record,
    remaining_records: &mut impl Iterator<Item=Record>
) -> eyre::Result<Option<AccumulatedStepTimings>> {
    let step_path = step_new_record.span_path();

    let mut accumulator = TimingAccumulator::new();
    accumulator.enter_span(step_path.clone(), step_new_record.timestamp().clone())?;

    let step_index = step_new_record.span()
        .and_then(|span| span.fields().pointer("/step_index"))
        .and_then(|value| value.as_u64())
        .ok_or_else(|| eyre!("step span does not have step_index field"))?;

    while let Some(record) = remaining_records.next() {
        if record.thread_id() == step_new_record.thread_id() {
            if let Some(span) = record.span() {
                match record.kind() {
                    SpanEnter => {
                        accumulator.enter_span(record.span_path(),
                                               record.timestamp().clone())?;
                    },
                    SpanExit => {
                        // TODO: use a stack to verify that open/close events are consistent?
                        let mut span_path = record.span_path();
                        // Close events don't report the current span anymore,
                        // so we need to add this to get a path consistent with the
                        // "new" event
                        span_path.push_span_name(span.name().to_string());
                        let is_step_span_path = span_path == step_path;
                        accumulator.exit_span(span_path,
                                              record.timestamp().clone())?;
                        if span.name() == "step"
                            && record.target() == "dynamecs_app"
                            && is_step_span_path {
                            break;
                        }
                    }
                    _ => {}
                }
            }
        }
    }

    if accumulator.has_active_spans() {
        // If there are active spans, then the step is not yet complete,
        // so we do not want to include it in accumulation
        // (would lead to inconsistent time between parent and children)
        Ok(None)
    } else {
        Ok(Some(AccumulatedStepTimings {
            timings: AccumulatedTimings { span_stats: accumulator.collect_completed_statistics() },
            step_index
        }))
    }
}

#[derive(Debug)]
struct TimingAccumulator {
    completed_statistics: HashMap<SpanPath, DirectStats>,
    enter_timestamps: HashMap<SpanPath, OffsetDateTime>,
}

impl TimingAccumulator {
    pub fn new() -> Self {
        Self { completed_statistics: Default::default(), enter_timestamps: Default::default() }
    }

    pub fn enter_span(&mut self, path: SpanPath, timestamp: OffsetDateTime) -> eyre::Result<()> {
        if self.enter_timestamps.insert(path, timestamp).is_some() {
            return Err(eyre!("tried to create new span that is already active (not closed)"));
        }
        Ok(())
    }

    pub fn exit_span(&mut self, path: SpanPath, timestamp_close: OffsetDateTime) -> eyre::Result<()> {
        let timestamp_enter = self.enter_timestamps.remove(&path)
            .ok_or_else(|| eyre!("found close event for span that is not currently active. Span path: {path}"))?;
        let span_duration: Duration = (timestamp_close - timestamp_enter).unsigned_abs();
        let accumulated_stats = self.completed_statistics.entry(path)
            .or_default();
        accumulated_stats.combine_mut(&DirectStats::from_single_duration(span_duration));
        Ok(())
    }

    pub fn has_active_spans(&self) -> bool {
        !self.enter_timestamps.is_empty()
    }

    pub fn collect_completed_statistics(self) -> HashMap<SpanPath, DirectStats> {
        self.completed_statistics
    }
}

