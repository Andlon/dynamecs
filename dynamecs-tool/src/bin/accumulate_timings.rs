use std::collections::HashMap;
use std::error::Error;
use std::{fmt, slice};
use std::io::{stdout, Write};
use std::time::Duration;
use tabwriter::TabWriter;
use dynamecs_analyze::iterate_records;
use dynamecs_analyze::timing::{accumulate_timings, SpanTiming2};

fn main() -> Result<(), Box<dyn Error>> {
    if let Some(arg) = std::env::args().skip(1).next() {
        let records: Vec<_> = iterate_records(&arg)?.collect::<Result<_, _>>()?;
        let timings = accumulate_timings(&records)?;

        // Map (parent, name) to durations, so that we can look up later
        let timings_map = compute_timings_map(&timings);
        let self_durations_map = compute_self_durations(&timings, &timings_map);

        let stdout = std::io::stdout().lock();
        let mut writer = TabWriter::new(stdout);
        writeln!(writer, "Duration\tSelf duration\tSelf %\t% of root\t% of parent\t Span")?;
        for timing in timings {
            let indent = {
                let pre_indent_num = timing.parent().len().saturating_sub(1);
                let mut indent = "│   ".repeat(pre_indent_num);
                if timing.parent().len() > 0 {
                    indent.push_str("├── ");
                }
                indent
            };
            let duration = format_duration(timing.duration());
            let root_ancestor = timing.parent().first();
            let root_duration = root_ancestor.and_then(|root| timings_map.get(slice::from_ref(root)).copied());
            let parent_duration = timings_map.get(timing.parent()).copied();
            let self_duration = self_durations_map.get(timing.span_path()).copied()
                .expect("Self duration must always be present");
            let self_duration_str = format_duration(self_duration);
            writeln!(writer, "{duration}\t{self_duration}\t{self_proportion}\t{root_proportion}\t{parent_proportion} \t {indent}{name}",
                     name = timing.name(),
                     self_duration = self_duration_str,
                     self_proportion = format_proportion(self_duration, Some(timing.duration())),
                     root_proportion = format_proportion(timing.duration(), root_duration),
                     parent_proportion = format_proportion(timing.duration(), parent_duration))?;
        }
        writer.flush()?;
    } else {
        eprintln!("No path to log file provided, exiting...");
    }
    Ok(())
}

fn compute_timings_map(timings: &[SpanTiming2]) -> HashMap<Vec<String>, Duration> {
    // Map (parent, name) to durations, so that we can look up later
    timings.iter()
        .map(|timing| {
            let mut path = timing.parent().to_vec();
            path.push(timing.name().to_string());
            (path, timing.duration())
        })
        .collect()
}

fn compute_self_durations(
    timings: &[SpanTiming2],
    timings_map: &HashMap<Vec<String>, Duration>
) -> HashMap<Vec<String>, Duration> {
    let mut self_durations_map = timings_map.clone();
    for timing in timings {
        if let Some(mut parent_self_duration) = self_durations_map.get_mut(timing.parent()) {
            *parent_self_duration -= timing.duration();
        }
    }
    self_durations_map
}

fn format_proportion(duration: Duration, other_duration: Option<Duration>) -> String {
    let proportion = other_duration.map(|relative_to| duration.as_secs_f64() / relative_to.as_secs_f64());
    proportion.map(|p| format!("{:5.1} %", p * 100.0))
        .unwrap_or_else(|| "N/A".to_string())
}

fn format_duration(duration: Duration) -> String {
    let s = duration.as_secs_f64();
    if s < 1e-7 {
        format!("{:7.2} ns", s * 1e9)
    } else if s < 1e-4 {
        format!("{:7.2} µs", s * 1e6)
    } else if s < 1e-1 {
        format!("{:7.2} ms", s * 1e3)
    } else {
        format!("{:7.2e}  s", s)
    }
}