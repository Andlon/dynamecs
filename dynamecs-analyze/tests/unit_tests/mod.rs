use serde_json::json;
use time::format_description::well_known::Iso8601;
use time::OffsetDateTime;
use dynamecs_analyze::{iterate_records_from_reader, Level, Record, RecordKind};
use dynamecs_analyze::timing::accumulate_timings;

/// Helper macro for succinctly creating a span path from a list of literals.
macro_rules! span_path {
    ($($strings:expr),*) => {
        SpanPath::new(vec![$($strings.to_string()),*])
    }
}

mod span_path;
mod span_tree;
mod timing;

#[test]
fn test_basic_records_iteration() {
    let log_data = r###"
        {"timestamp":"2023-03-29T12:48:50.213348Z","level":"TRACE","fields":{"message":"enter"},"target":"dynsys::backward_euler","span":{"name":"Backward Euler IP assemble"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"Backward Euler IP assemble"}]}
        {"timestamp":"2023-03-29T12:48:51.440914Z","level":"INFO","fields":{"message":"exit"},"target":"dynsys::backward_euler","span":{"name":"hessian"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"Backward Euler IP assemble"}]}
        {"timestamp":"2023-03-29T12:48:51.440972Z","level":"TRACE","fields":{"message":"exit"},"target":"dynsys::backward_euler","span":{"name":"Backward Euler IP assemble"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"}]}
        {"timestamp":"2023-03-29T12:48:51.441519Z","level":"DEBUG","fields":{"message":"enter"},"target":"dynsys::backward_euler","span":{"name":"solve_linear_system"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"solve_linear_system"}]}
    "###;
    let records: Vec<Record> = iterate_records_from_reader(log_data.as_bytes())
        .collect::<eyre::Result<_>>()
        .unwrap();

    assert_eq!(records.len(), 4);

    {
        let record = &records[0];
        assert_eq!(record.level(), Level::Trace);
        assert_eq!(record.target(), "dynsys::backward_euler");
        assert_eq!(record.kind(), RecordKind::SpanEnter);
        assert_eq!(record.message(), Some("enter"));
        assert_eq!(record.timestamp(), &OffsetDateTime::parse("2023-03-29T12:48:50.213348Z", &Iso8601::DEFAULT).unwrap());
        assert_eq!(record.span().unwrap().name(), "Backward Euler IP assemble");
        assert_eq!(record.span().unwrap().fields(), &json! {{
            "name": "Backward Euler IP assemble",
        }});
        assert_eq!(record.spans().unwrap().len(), 6);
        assert_eq!(record.spans().unwrap()[0].name(), "run");
        assert_eq!(record.spans().unwrap()[0].fields(), &json! {{
            "name": "run"
        }});
        assert_eq!(record.spans().unwrap()[1].name(), "step");
        assert_eq!(record.spans().unwrap()[1].fields(), &json! {{
            "name": "step",
            "step_index": 16
        }});
        assert_eq!(record.spans().unwrap()[2].name(), "Backward Euler");
        assert_eq!(record.spans().unwrap()[2].fields(), &json! {{
            "name": "Backward Euler"
        }});
        assert_eq!(record.spans().unwrap()[3].name(), "Backward Euler");
        assert_eq!(record.spans().unwrap()[3].fields(), &json! {{
            "name": "Backward Euler"
        }});
        assert_eq!(record.spans().unwrap()[4].name(), "Newton iteration");
        assert_eq!(record.spans().unwrap()[4].fields(), &json! {{
            "name": "Newton iteration",
            "hessian_mod": "NoModification",
            "k": 8,
        }});
        assert_eq!(record.spans().unwrap()[5].name(), "Backward Euler IP assemble");
        assert_eq!(record.spans().unwrap()[5].fields(), &json! {{
            "name": "Backward Euler IP assemble",
        }});
    }

    {
        let record = &records[1];
        assert_eq!(record.level(), Level::Info);
        assert_eq!(record.target(), "dynsys::backward_euler");
        assert_eq!(record.kind(), RecordKind::SpanExit);
        assert_eq!(record.message(), Some("exit"));
        assert_eq!(record.timestamp(), &OffsetDateTime::parse("2023-03-29T12:48:51.440914Z", &Iso8601::DEFAULT).unwrap());
        assert_eq!(record.span().unwrap().name(), "hessian");
        assert_eq!(record.span().unwrap().fields(), &json! {{
            "name": "hessian",
        }});
        assert_eq!(record.spans().unwrap().len(), 6);

        // TODO: Test the rest
    }
}

#[test]
fn test_basic_timing_collection() {
    let log_data = r###"
{"timestamp":"2023-03-29T18:16:50.073142Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.073564Z","level":"INFO","fields":{"message":"Starting simulation of scenario \"basic\""},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.073929Z","level":"INFO","fields":{"message":"No time step configured. Using default dt = 0.016666666666666666"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.074164Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.074370Z","level":"DEBUG","fields":{"message":"Running post-systems for initial state"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.074565Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.074771Z","level":"DEBUG","fields":{"message":"some debug info","number":42},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.074911Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075098Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.075291Z","level":"DEBUG","fields":{"message":"some other debug info","digits":17},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.075428Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075745Z","level":"INFO","fields":{"message":"Starting step 0 at simulation time 0.00000 (dt = 1.66667e-2)"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075945Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076217Z","level":"DEBUG","fields":{"message":"some debug info","number":42},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076413Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.076594Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076780Z","level":"DEBUG","fields":{"message":"some other debug info","digits":17},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076916Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.077044Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077210Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"step_index":1,"name":"step"},"spans":[{"name":"run"},{"step_index":1,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.077323Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"step_index":1,"name":"step"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077568Z","level":"INFO","fields":{"message":"Simulation ended"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077681Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"run"},"spans":[]}
    "###;
    let records: Vec<_> = iterate_records_from_reader(log_data.as_bytes())
        .collect::<eyre::Result<_>>()
        .unwrap();
    let timings = accumulate_timings(&records).unwrap();
    insta::assert_debug_snapshot!(timings);
}