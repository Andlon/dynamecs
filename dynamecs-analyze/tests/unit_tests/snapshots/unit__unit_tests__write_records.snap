---
source: dynamecs-analyze/tests/unit_tests/mod.rs
expression: records_string
---
{"timestamp":"2023-02-22T08:00:00+02:00","level":"INFO","fields":{"message":"msg0"},"target":"a","threadId":"0"}
{"timestamp":"2023-02-22T08:00:01+02:00","level":"TRACE","fields":{"message":"msg1"},"target":"a","threadId":"0"}
{"timestamp":"2023-02-22T08:00:02+02:00","level":"INFO","fields":{"message":"enter"},"target":"b","span":{"name":"span1"},"spans":[{"name":"span1"}],"threadId":"0"}
{"timestamp":"2023-02-22T08:00:03+02:00","level":"DEBUG","fields":{"message":"message2"},"target":"b","span":{"name":"span1"},"spans":[{"name":"span1"}],"threadId":"1"}
{"timestamp":"2023-02-22T08:00:04+02:00","level":"WARN","fields":{"field1":4,"field2":"value2"},"target":"b","span":{"name":"span1"},"spans":[{"name":"span1"}],"threadId":"0"}
{"timestamp":"2023-02-22T08:00:05+02:00","level":"INFO","fields":{"message":"exit"},"target":"b","span":{"name":"span1"},"threadId":"0"}

